test/path_tokenizer.h

```diff
@@ -0,0 +1,52 @@
+//
+// Created by wkl04 on 12-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+#define MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+
+inline void mqtt_path_tokenizer(MQTT_NS::string_view str, std::function< bool(MQTT_NS::string_view str) > const &write)
+{
+    auto start = str.begin();
+    auto end = str.end();
+
+    auto current = start;
+    while(current != str.end()) {
+        if(*current == '/') {
+            write(MQTT_NS::string_view(start, std::distance(start, current)));
+            ++current;
+            start = current;
+        } else {
+            ++current;
+        }
+    }
+
+    write(MQTT_NS::string_view(start, std::distance(start, current)));
+}
+
+// See if the topic is valid (does not contain # or +)
+inline bool mqtt_valid_topic(MQTT_NS::string_view topic)
+{
+    if(topic.empty()) {
+        return false;
+    }
+
+    if(topic[0] == '$') {
+        return false;
+    }
+
+    return (topic.find_first_of("#+") == std::string::npos);
+}
+
+// See if the subscription is valid
+inline bool mqtt_valid_subscription(MQTT_NS::string_view subscription)
+{
+    if(subscription.empty()) {
+        return false;
+    }
+
+    // If hash position is found, it should be the last character
+    return (subscription.find('#') == std::string::npos) || (subscription.back() == '#');
```

Comment:
I don't think this is sufficient.

You're checking that either there is *no* #, or there is *a* # as the last character.

I think "#asdfasdfsadfasdf#" would return true.